File: rholang/src/test/scala/coop/rchain/rholang/interpreter/DeployDataSpec.scala
package coop.rchain.rholang.interpreter

import cats.effect.IO
import cats.effect.concurrent.Ref
import coop.rchain.metrics.{Metrics, NoopSpan, Span}
import coop.rchain.shared.{Base16, Log}
import monix.eval.Task
import monix.execution.Scheduler.Implicits.global
import coop.rchain.crypto.PublicKey
import coop.rchain.crypto.hash.Blake2b512Random
import coop.rchain.models.Expr.ExprInstance.{GInt, GString}
import coop.rchain.models.rholang.implicits._
import coop.rchain.models._
import coop.rchain.rholang.interpreter.RhoRuntime.RhoISpace
import coop.rchain.rholang.interpreter.InterpreterUtil._
import coop.rchain.rholang.interpreter.SystemProcesses.DeployData
import coop.rchain.rholang.syntax._
import coop.rchain.rspace.syntax._
import coop.rchain.models.syntax._
import com.google.protobuf.ByteString
import org.scalatest._
import org.scalatest.Matchers._

class DeployDataSpec extends FlatSpec with Matchers {

  implicit val logF: Log[Task]            = new Log.NOPLog[Task]
  implicit val noopMetrics: Metrics[Task] = new Metrics.MetricsNOP[Task]
  implicit val noopSpan: Span[Task]       = NoopSpan[Task]()

  val errorHandler                    = Ref.unsafe[IO, Vector[Throwable]](Vector.empty)
  implicit val rand: Blake2b512Random = Blake2b512Random(Array.empty[Byte])

  "rho:deploy:data system channel" should "return timestamp, deployerId and deployId" in {
    val contract =
      """
        |new deployData(`rho:deploy:data`) in {
        |  deployData!(0)
        |}
      """.stripMargin

    val timestamp = 123L;
    val key       = PublicKey(Base16.unsafeDecode("abcd"));
    val sig       = Base16.unsafeDecode("1234").toByteString;

    TestDeployDataFixture.test(
      contract,
      DeployData(timestamp, key, sig),
      List(
        Par(exprs = Seq(Expr(GInt(timestamp)))),
        Par(
          unforgeables = Seq(
            GUnforgeable(
              GUnforgeable.UnfInstance.GDeployerIdBody(GDeployerId(key.bytes.toByteString))
            )
          )
        ),
        Par(
          unforgeables = Seq(
            GUnforgeable(
              GUnforgeable.UnfInstance.GDeployIdBody(GDeployId(sig))
            )
          )
        )
      )
    )
  }
}

object TestDeployDataFixture {
  import coop.rchain.rholang.Resources._

  def test(contract: String, deployData: DeployData, expected: List[Par]): Unit = {
    implicit val logF: Log[Task]            = new Log.NOPLog[Task]
    implicit val noopMetrics: Metrics[Task] = new Metrics.MetricsNOP[Task]
    implicit val noopSpan: Span[Task]       = NoopSpan[Task]()

    val result = mkRuntime[Task]("rho-deploy-data-test")
      .use { rhoRuntime =>
        for {
          _ <- rhoRuntime.setDeployData(deployData)
          _ <- evaluate[Task](
                rhoRuntime,
                contract
              )
          data <- rhoRuntime.getData(GInt(0L))
        } yield {
          if (data.nonEmpty) data.head.a.pars
          else Nil
        }
      }
      .runSyncUnsafe()

    result.toSeq shouldEqual expected
  }
}

File: rholang/tests/deploy_data_spec.rs
// See rholang/src/test/scala/coop/rchain/rholang/interpreter/DeployDataSpec.scala

use crypto::rust::public_key::PublicKey;
use models::rhoapi::{
    expr::ExprInstance, g_unforgeable::UnfInstance, BindPattern, Expr, GDeployId, GDeployerId,
    GUnforgeable, ListParWithRandom, Par, TaggedContinuation,
};
use rholang::rust::interpreter::{
    external_services::ExternalServices,
    matcher::r#match::Matcher,
    rho_runtime::{create_rho_runtime, RhoRuntime},
    system_processes::DeployData,
};
use rspace_plus_plus::rspace::{
    rspace::RSpace,
    shared::{
        in_mem_store_manager::InMemoryStoreManager, key_value_store_manager::KeyValueStoreManager,
    },
};
use std::sync::Arc;

#[tokio::test]
async fn rho_deploy_data_system_channel_should_return_timestamp_deployer_id_and_deploy_id() {
    let contract = r#"
        new deployData(`rho:deploy:data`) in {
          deployData!(0)
        }
    "#;

    let timestamp = 123i64;
    let key = PublicKey::from_bytes(&hex::decode("abcd").unwrap());
    let sig = hex::decode("1234").unwrap();

    let deploy_data = DeployData {
        timestamp,
        deployer_id: key.clone(),
        deploy_id: sig.clone(),
    };

    let expected = vec![
        Par::default().with_exprs(vec![Expr {
            expr_instance: Some(ExprInstance::GInt(timestamp)),
        }]),
        Par::default().with_unforgeables(vec![GUnforgeable {
            unf_instance: Some(UnfInstance::GDeployerIdBody(GDeployerId {
                public_key: key.bytes.as_ref().to_vec(),
            })),
        }]),
        Par::default().with_unforgeables(vec![GUnforgeable {
            unf_instance: Some(UnfInstance::GDeployIdBody(GDeployId { sig })),
        }]),
    ];

    TestDeployDataFixture::test(contract, deploy_data, expected).await;
}

struct TestDeployDataFixture;

impl TestDeployDataFixture {
    async fn test(contract: &str, deploy_data: DeployData, expected: Vec<Par>) {
        let mut kvm = InMemoryStoreManager::new();
        let store = kvm.r_space_stores().await.unwrap();
        let space: RSpace<Par, BindPattern, ListParWithRandom, TaggedContinuation> =
            RSpace::create(store, Arc::new(Box::new(Matcher))).unwrap();

        let mut runtime = create_rho_runtime(
            space,
            Par::default(),
            true,
            &mut Vec::new(),
            ExternalServices::noop(),
        )
        .await;

        runtime.set_deploy_data(deploy_data).await;

        runtime.evaluate_with_term(contract).await.unwrap();

        let channel = Par::default().with_exprs(vec![Expr {
            expr_instance: Some(ExprInstance::GInt(0)),
        }]);

        let data = runtime.get_data(&channel);

        let result: Vec<Par> = if !data.is_empty() {
            data.into_iter().flat_map(|datum| datum.a.pars).collect()
        } else {
            Vec::new()
        };

        assert_eq!(result, expected);
    }
}
